net_pkt buffer management and API changes

tbursztyka opened the pull request:
This is in a early stage, some stuff will change in these patches still.

In order not to break anything (besides some test due to ROM/RAM being bloated for the time being), the API is made in parallel to the existing one (thus the bloat). Currently, as the patches can tell you, only part of ipv4/Ethernet are moved towards this new API. The good thing with this approach is that old API works on packet generated with the new API and vice versa so changes can be made iteratively.

Commit:
test/net: Quick sample explaining/testing the new net_pkt buffer API

This goes through basic allocation, atomic read/write, data
getter/setter and cloning/copying.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Add new allocators

These struct net_pkt allocators will give the possibility to allocate at
once the net_pkt and the buffer associated with, taking care of the
header space and MTU relevantly.

This enables to use the variable length allocator from net_buf. However,
it is not yet the default and is set as experimental.

As it is provided in parallel to existing allocators, it has to keep a
slab per-direction and thus a pointer in net_pkt, as well as appdata,
appdatalen etc... Resulting in "bloating" net_pkt. This will be solved
when, finally, former allocators will be removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Add an API to read/write into the buffer in a simpler way

Adding a cursor into net_pkt. This is used to read/write data in a much
simpler way, for pre-allocated buffers in net_pkt. This avoids API users
to deal with net_buf below directly.

However, to be used - as for the new allocators - it will require deep
net stack core and API changes.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Add dummy getter/setters for IPv4/IPv6 attributes

All as static inline functions to let the compiler check the types
etc... And use ARG_UNUSED() always where relevant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Add generic get/set data functions

These ones would support linearizing non-contiguous area, however
requiring a bit more complex type as an "accessor".

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/udp: Add new API to create/finalize UDP headers

This API is meant to work with pre-allocated net_pkt.
It assumes net_pkt's buffer cursor is at the right position where to
create the UDP header. Once done, the cursor will be placed right after
the newly created UDP header.

Finalizing assumes the same.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv4: Add new API to create/finalize IPv4 headers.

This API is meant to work with pre-allocated net_pkt.
It assumes net_pkt's buffer cursor is at the right position where to
create the IPv4 header. Once done, the cursor will be placed right
after the newly created IPv4 header.

Finalizing assumes the same.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv4: Switch echo request to new net pkt allocator and API

This is the easiest piece to change using the new API, and can be used
as showing how this API can be used.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ethernet: Use new net_pkt API for sending and receiving

Use the new API where relevant. Only sam_gmac is left aside for now.

This simplifies a lot the code as the caller should only care about
allocating net_pkt and its buffer once, and thus will not need to mess
with "frags" etc...

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/dhcpv4: Switch to new net_pkt API

Use new allocator and function to r/w dhcp messages.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented:
@tbursztyka, how this relates to https://github.com/zephyrproject-rtos/zephyr/pull/11374 ?

tbursztyka commented:
@pfalcon There are 2 relations and 1 big reason
1 indirect relation: In current buffer management, it's up to L3 to allocate L2 buffer space. That could be fine if all bearers had a fixed amount of space required. But it's not. 802.15.4 has a variable buffer space requirement, depending on many parameters which I am not listing here, such buffer space cannot be predicted until you know exactly where the packet is meant to go. Which information is known only after IPv6 ND process. And thus forces you to update this space (it might shrink or grow).
So properly splitting L2 to L3 fixes this. Once a packet reaches L2, L3 has all the info where it is supposed to go (it ran ND already etc..). So L2 will be able to compute, once, the exact amount of buffer required and do its job. For this new buffer management, it was necessary to fix this to simplify the whole process a lot. L3 has to take care about L3 only.
1 direct relation: the compatibility of old vs new API on complete packet (received or freshly created ones for tx). Due to this ll buffer space being allocated on L3, and because then L3 did not care about it, it required to "hide" this space via net_buf_pull(). But then on l2 and in each and every drivers, it of course needed to retrieve that space (via the pkt->frags->data - net_pkt_ll_reserve(pkt). And only the first "frag" had its reserved space used relevantly, so it need to be handled specifically (some ethernet drivers where having comments describing this). And that would have been an issue for using new API net_pkt_read/net_pkt_write or else, since these do not know if you need to access this "hidden space" or not.

And 1 big and obvious reason: Optimizing buffer usage. This ll reserve thingy was thus mixed in L3, which architecturally speaking was bad already. But it was even worse in net_buf allocation mechanism: _all_ struct net_buf had it's size shrunk by the ll_reserve amount. So if your frag size is 128bytes, used to built a big ethernet frame you would loose 14 bytes on all fragments but the first one.

Side note: the "frag" size was also, due to this weird L3/L2 mix-up tight to the bearer which had the smallest MTU. And that was again a weird thing, since buffer management is memory management: it should not have any relationship with what bearers is used or else. Theoretically you should be able to go as low as 1 byte per-net_buf and all should work just the same. (That would not make any sense in practice obviously, having a descriptor such as a struct net_buf n times bigger than the buffer it points to)

Last but not least: doing that split, offers other optimizations possibilities like in ethernet: you could have a dedicated net_buf for the ethernet header part, allocated on stack, with just 14 bytes as buffer etc... On 15.4 it will be also possible then to avoid modifying the original packet (6lo and 15.4 fragmentation were doing so before), and thus avoid - when TCP is used - to copy the whole packet in TCP for possible needed retries etc...

Also, an additional reason of doing this PR is to hide all the ugly buffer details to most of the code. We have these ugly stuff everywhere:
frag = pkt->frags;
while (frags) {
     ...
     frag = frag->frags;
}

Instead we should not do any of it anywhere, you just write/read stuff from your packet. (and in case of network headers, you just access a header pointer directly).

Commit:
net: Adapt checksum calculation to new net_pkt API

Let's just use the packet cursor relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on samples/net/ws_echo_server/src/main.c:
```diff
@@ -97,12 +97,6 @@ struct net_pkt *build_reply_pkt(const char *name,
 			net_buf_push(frag, net_buf_headroom(frag));
```
Same with this one.

jukkar commented on include/net/net_pkt.h:
```diff
@@ -2007,6 +2010,156 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
+
+struct net_pkt *net_pkt_allocate_debug(s32_t timeout,
+				       const char *caller, int line);
+#define net_pkt_allocate(_timeout)					\
```
What about making the function name a bit shorter so I propose we change this to `net_pkt_alloc()`

tbursztyka replied:
true. I had that name at first, I don't know why I went for the plain allocate word.

jukkar commented on include/net/net_pkt.h:
```diff
@@ -2007,6 +2010,156 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
+ *
+ * @return a pointer to a newly allocated net_pkt on success, NULL otherwise.
+ */
+struct net_pkt *net_pkt_allocate(s32_t timeout);
```
Shouldn't we call this net_pkt_tx_allocate() in order to follow the same syntax as the rx allocator one?

tbursztyka replied:
the rx stuff is an in-between name. Idea would be to have only 1 allocator for all pkt (thus one slab, etc...). However, we will still be able to know (for stats, net mem in shell etc...) if a pkt was allocated for tx or rx, there will be a bit in net_pkt to give that info and so on (net_recv_data will set it for rx, tx will be default). That's not going to be part of that PR though,

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -2152,6 +2164,482 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+				size_t existing)
+{
+	sa_family_t family = net_pkt_family(pkt);
+	size_t max_len;
```
what about getting max_len like this
```
	size_t max_len = net_if_get_mtu(net_pkt_iface(pkt));

	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
		max_len = max(max_len, NET_IPV6_MTU);
	} else if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
		max_len = max(max_len, NET_IPV4_MTU);
	} 
```

then we do not need the AF_UNSPEC branch at all.

tbursztyka replied:
indeed

jukkar commented on subsys/net/ip/Kconfig:
```diff
@@ -472,6 +472,19 @@ config NET_BUF_DATA_POOL_SIZE
 	 This value tell what is the size of the memory pool where each
 	 network buffer are allocated from.
 
+config NET_HEADERS_ALWAYS_CONTIGUOUS
```
Offering this kind of option is imho dangerous, someone will misuse it anyway. How much memory we are saving with this, what does ram_report say if we enable/disable it?

tbursztyka replied:
for udp/ipv4 setup, even on 128bytes frag size this will be just fine option (it is dangerous at the same level as wrongly teawking frag size: user has to know what he is doing). It saves stack actually.

jukkar replied:
Sure, it probably save something. Just wondering do you have numbers about the savings.

jukkar commented on subsys/net/ip/ipv4.h:
```diff
@@ -52,4 +52,33 @@ struct net_pkt *net_ipv4_create(struct net_pkt *pkt,
+int net_ipv4_create_new(struct net_pkt *pkt,
+			const struct in_addr *src,
+			const struct in_addr *dst,
+			u8_t next_header_proto);
```
Why do we need to have next_header_proto parameter here and in finalize function, it looks like one should be enough.

tbursztyka replied:
you mean we would inform ipv4 header about the protocol only at finalize? why not indeed.

jukkar commented on include/net/net_pkt.h:
```diff
@@ -2160,6 +2184,228 @@ struct net_pkt *net_pkt_rx_allocate_with_buffer(struct net_if *iface,
+int net_pkt_read_new(struct net_pkt *pkt, void *data, size_t length);
+
+/* Read u8_t data data a net_pkt */
+static inline int net_pkt_read_new_u8(struct net_pkt *pkt, u8_t *data)
```
I suggest we change the name to `net_pkt_read_u8_new()` in order to be consistent. Same with the other value reader helpers.

tbursztyka replied:
will be easier to remove the _new when necessary yes

jukkar reviewed (changes requested):
Lot of commits but fortunately most of them were small and readable.

pfalcon commented:
@tbursztyka: I appreciate the detailed write up in https://github.com/zephyrproject-rtos/zephyr/pull/11775#issuecomment-443632138.

My original question re: relation to https://github.com/zephyrproject-rtos/zephyr/pull/11374 was simpler actually, to confirm that #11374 precedes this patch, instead of e.g. be superseded by this one, or a case when they largely unrelated.  So, I just assume that #11374 is the next in queue, and not anything else.

tbursztyka commented:
@pfalcon yes PR #11374 is the first in queue.

Commit:
net/core: Each and every received packet are being set to overwrite

This will avoid new API's writing functions to modify the packet's
content. For instance while checking its checksum etc...

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented:
In one of recent commits:

> c965644: net/core: Each and every received packet are bein set to overwrite

"is being"

Commit:
net/ipv4: Input headers are gathered through new API

As before, such header is meant to be in a contiguous area (beginning
of the buffer, only 20 bytes)

Opportunistically chaning the function name to net_ipv4_input() (all
will be create/finalize/input).

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv4: Input headers are gathered through new API

As before, such header is meant to be in a contiguous area.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv4: Add a new function to finalize the ICMPv4 packet

Function names will be normalized then by the couple create/finalize.
This one only sets the checksum.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv4: Rework relevant signatures to pass ipv4 header pointer

This is meant to remove the need for macro NET_IPV4_HDR(), since we
don't know in future if accessing the header that way will be valid.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv4: Switch echo request handler to new API

Unlike before, we allocate a new packet for the reply which is a
modified clone of the request.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv4: Remove useless create, get and set header functions

These are not needed anymore.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv4: Switch error message to new API

Logic does not change much, besides using new create/finalize functions
etc...

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/core: Set cursor relevantly

Newly received pkt can get their cursor intialized at net_recv_data()
(most of the time, drivers won't mangle with the content before calling
that function).

Right after l2 (net_if_recv_data()) parsing as well. L2s pull the
starting buffer after ll header. Instead of letting L2s updating the
cursor, it's simpler to reinitialize it directly after such parsing.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Create new send sendto functions using new net_pkt API

It's not anymore up to user to provide the pkt. Context will build the
packet according to its metadata and provided buffer and length.

It currently supports only IPv4 and UDP.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/udp: Add an input function

This will be the place where UDP header is parsed: its checksum verified
and src/dst ports grabbed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/tcp: Add an input function for pre-validation

This will quickly verify the header, its checksum and grab src/dst
ports.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/connection: Modify input function

Verifying udp/tcp checksum should be done before calling this function.
Also, up to ipv4 and ipv6 to provide a pointer to their respective
header.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv4: Rework input function

Use udp/tcp input functions relevantly, and call net_conn_input
afterwards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/sockets: Use the new send/sendto from net_context

It's now up to net_context to build the net_pkt and send it.
This will become the default.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/net: Avoid using net_udp_get_chksum()

checksum_offload is now the only one using that function, let's get rid
of such usage.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/udp: Remove now useless net_udp_get_chksum function

It's not used anywhere.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Add new API to create/finalize IPv6 headers.

This API is meant to work with pre-allocated net_pkt.
It assumes net_pkt's buffer cursor is at the right position where to
create the IPv6 header. Once done, the cursor will be placed right
after the newly created IPv6 header.

Finalizing assumes the same.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv6: Switch echo request to new net pkt allocator and API

This is pretty much the same as in ICMPv4. Actually, when it comes to
the ICMP header, it could probably be put in a common places for both
ICMPv4 and ICMPv6.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv6: Add a new function to finalize the ICMPv6 packet

Function names will be normalized then by the couple create/finalize.
This one only sets the checksum.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Rework input function

Use the new net_pkt API to proceed through IPv6 header and all the
extension header as well.

Use udp/tcp input functions relevantly, and call net_conn_input
afterwards.

Note: This commit temporarly disable IPv6 fragmentation support
in the code directly. Which support will be re-enabled afterwards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented:
Commit:
>  net/ip: Protocol family do not have to follow any other OS's value

> It is the macro name that matters, not its value. Here, that will help
to save 1 bit in struct net_pkt later on.

Those are not "any other OS's value", those are official IANA numbers for protocols. You can skip talking "ipv4 protocol" and "ipv6 protocols", and say "protocol 2" and "protocol 10", and people are supposed to understand you ;-). Especially if it's your application talking (e.g. in log output).

pfalcon commented on tests/net/context/src/main.c:
```diff
@@ -88,7 +88,7 @@ static void net_ctx_get_fail(void)
 		      "Invalid context protocol test failed");
 
-	ret = net_context_get(1, SOCK_DGRAM, IPPROTO_UDP, &context);
+	ret = net_context_get(4, SOCK_DGRAM, IPPROTO_UDP, &context);
```
Why not use clearly invalid number here, say 1234?

pfalcon commented:
> net/pkt: Add dummy getter/setters for IPv4/IPv6 attributes 

This can (should) be submitted on its own and merged right away.

tbursztyka commented:
@pfalcon Can you give an actual reference about that? I was looking at RFCs and all but could not pin point a place where these macros have specified fixed values.

pfalcon commented:
>> Those are not "any other OS's value", those are official IANA numbers for protocols.

> @pfalcon  Can you give an actual reference about that? I was looking at RFCs and  all but could not pin point a place where these macros have specified  fixed values.

Interestingly, looking up the actual reference, it disproves what I said. There're official IANA numbers, but they happen to match not what we had, but what you introduce in your patch: https://www.iana.org/assignments/address-family-numbers/address-family-numbers.xhtml (Note: this is named "address families", but numbers from those are known to match "protocol families". Except that "known" from Linux, again.)

So yes, "2" and "10" were number from a specific OS, Linux. If they don't come from IANA, they apparently come from iBCS2. And my practical point is that for higher levels of stacks, like BSD Socket API, I'd prefer PF_INET to have value 2, PF_INET6 - 10. We can use other numbers in lower stack, but they should be named differently then.

(All that again if just my PoV/suggestion.)

pfalcon commented:
FreeBSD has AF_INET of 2, AF_INET6 of 28: https://github.com/freebsd/freebsd/blob/master/sys/sys/socket.h . Well, what can I say. Let's make more of a zoo, why not.

tbursztyka commented:
Yep, exactly what I found out: this numbering does not seem to be strictly specified. I guess macro names are the important part here and that of course we we'll follow strictly.

pfalcon commented on include/net/net_pkt.h:
```diff
@@ -1998,6 +2001,158 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
+					     s32_t timeout,
+					     const char *caller,
+					     int line);
+#define net_pkt_alloc_on_iface_(_iface, _timeout)			\
```
Is trailing underscore in `net_pkt_alloc_on_iface_` intended?

tbursztyka replied:
no

pfalcon commented on include/net/net_pkt.h:
```diff
@@ -1998,6 +2001,158 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
+ * @brief Allocate an initialized net_pkt
+ *
+ * Note: for the time being, 2 pools are used. One for TX and one for RX.
+ *       This allocater has to be used for TX.
```
So, why it's not named `net_pkt_tc_alloc`?

tbursztyka replied:
This "rx/tx pools" is not a good idea, it was a bandage on a wooden leg when it first appeared (worst one being the net_context dedicated pool). I want to get rid of it. So here, the core itself will not use any "tx" wording anywhere. For the time being, only drivers will have the differentiation with "rx".

Then one I am done with the API, I'll get rid of the tx/rx pools, one bit in net_pkt will tell the direction. unset: tx, set: rx. and rx will be set in net_recv_data(), then in memory info we will still know on what path packet are allocated. And that's all, that will be enough for what this differentiation is really used (and no such differentiation has never been a proper fix for memory leak or else)

pfalcon replied:
So, later is later, in the meantime, this patch gives "net_pkt_rx_alloc" vs "net_pkt_alloc", which is weird.

And using packet pools for different purposes, for example for RX vs TX direction, is the best practice to write reliable applications, so it needs to be supported anyway.

tbursztyka replied:
> is the best practice to write reliable applications

I don't know how you ended up to such conclusion.
If either rx or tx pool are empty due to a leak, the result will be _exactly_ the same as with one pool: there is nothing that can be done to recover.
Now in case one wants to limit how greedy is the rx path over tx path or vice versa, that also can be done with one pool, or limiting the rx/tx fifos etc... there are various ways. But having 2 pools is not the unique solution for this. Plus, the downside of this 2+ pools scheme is also the code everywhere: you need to use the right allocator, thus you need dedicated functions, thus it eats up a bit more ram/rom etc etc... It's only more burden for no good reasons.

pfalcon commented on include/net/net_pkt.h:
```diff
@@ -1998,6 +2001,158 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
+struct net_pkt *net_pkt_rx_alloc(s32_t timeout);
+
+/**
+ * @brief Allocate a network packet for a specific network interface.
```
> for a specific network interface

And why do we need such a function, given that we didn't need it before? Common sense says that the only thing which would need to allocate a packer "for a specific network interface" is the driver for this interface, but the driver already knows how to allocate packet for *this* specific interface (e.g. from a RX pool of the driver, vs generic system pool).

pfalcon replied:
Ok, actually that's more related to "net_pkt_rx_alloc_on_iface". Already got confused by "net_pkt_alloc_on_iface" vs "net_pkt_rx_alloc_on_iface" ;-).

pfalcon commented on include/net/net_pkt.h:
```diff
@@ -1998,6 +2001,158 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
+ * @param proto   The IP protocol type (can be 0 for none).
+ * @param timeout Maximum time in milliseconds to wait for an allocation.
+ *
+ * @return 0 on success, negative errno code otherwise.
```
nd - how to access this new buffer allocated?

tbursztyka replied:
It's still allocated through the infamous frags attribute (I really don't like this name, it's mixes up fragment of memory vs fragment of a packet, which is a concept that is present in IP also... meh, I introduced a "buffer" attribute, in a union with frags. Which one will disappear). So current API works the same. But the new API introduced in next patch is far easier to use. (one does not have to care about looping on "frags", having its own offset, frag, etc...).

pfalcon commented on subsys/net/ip/Kconfig:
```diff
@@ -437,17 +437,40 @@ config NET_BUF_TX_COUNT
 	  header (sizeof(struct net_buf)) amount of data.
 
+choice
+	prompt "Network packet data allocator"
```
Apparently should be "Network packet data allocator *type*".

tbursztyka replied:
sure

Commit:
net/icmpv6: Rework relevant signatures to pass ipv6 header pointer

This is meant to remove the need for macro NET_IPV6_HDR(), since we
don't know in future if accessing the header that way will be valid.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv6: Switch echo request handler to new API

Unlike before, we allocate a new packet for the reply which is a
modified clone of the request.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented on subsys/net/ip/net_pkt.c:
```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+	max_len -= existing;
+	max_len = min(size, max_len);
+
+	return min(size, max_len);
```
Duplicate call to max(), see line above.

pfalcon commented on subsys/net/ip/net_pkt.c:
```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+	return min(size, max_len);
+}
+
+static size_t pkt_headers_buffer_length(struct net_pkt *pkt,
```
Log message below says this returns *estimation* of headers len, why that's not mirrored in the name?

tbursztyka replied:
Sure pkt_estitame_headers_length() then

pfalcon commented on subsys/net/ip/net_pkt.c:
```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+
+static size_t pkt_get_size(struct net_pkt *pkt)
+{
+	struct net_buf *buf = pkt->buffer;
```
I think this member should be called buffer**s** after all.

tbursztyka replied:
no, because it represents the overall packet buffer. Fact that it might be made of one or multiple net_buf (in a list) is not relevant for the name here.

pfalcon replied:
> Fact that it might be made of one or multiple net_buf (in a list) is not relevant for the name here.

For us humans, it is. It wouldn't be relevant in a language with proper abstraction layer, like C++. But we have very thin, flat API here, where layout of data structures is part of API and thus naming should reflect that. Many people who just start with Zephyr IP stack would get an impression that there's a single buffer behind it and won't try to iterate on it. (Maybe your aim is to provide API which will discourage direct access to data structures, instead pushing access via abstracted method calls, but that's not how we had it (nor extra abstraction needed here, having thin abstractions for efficient code isn't bad)).

pfalcon commented on subsys/net/ip/net_pkt.c:
```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+	}
+
+	if (k_is_in_isr()) {
+		timeout = K_NO_WAIT;
```
Should be assert instead I guess?

tbursztyka replied:
no it's fine, it follows the same behavior as in net_buf API here.

pfalcon replied:
>>	if (k_is_in_isr()) {
		timeout = K_NO_WAIT;

> no it's fine, it follows the same behavior as in net_buf API here.

Yeah, but that doesn't mean that net_buf API is right doing that. Introducing new API gives a chance to tighten up API contract for the sake of more robust software. Because again, on one hand we put `(void)` in front of every memset() (MISRA), on the other, covering up such goofs a programmer may make.

pfalcon commented on subsys/net/ip/net_pkt.c:
```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+	/* Verifying existing buffer and take into account free space there */
+	alloc_len = pkt_get_size(pkt) - net_pkt_get_len(pkt);
+	if (!alloc_len) {
+		/* In case nothing is there, it will account for header
```
Is this condition really amounts to "nothing is there"? I'd think it's indeed "no free space", but header may be already there.

tbursztyka replied:
yep

pfalcon commented on subsys/net/ip/net_pkt.c:
```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+		timeout -= min(timeout, diff);
+	}
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
```
Any thinking of #ifdef'ing just the extra args, not duplicating same content twice, risking it getting out of sync and complicating studying code?

tbursztyka replied:
Using this is what has been done already. I don't like these if/else everywhere for the logging, but I will not address this problem here.
It would be nice to avoid all of these *_debug explicit declaration and all these if/else/endif

pfalcon commented on subsys/net/ip/net_pkt.c:
```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+#endif
+
+	if (pkt) {
+		net_pkt_set_iface(pkt, iface);
```
So, the whole purpose of these 2 extra functions to just set ->iface member of a pkt? Well, that's rather non-orthogonal API, I wonder if it really comes from tangible improvement with such design.

tbursztyka replied:
As I said in my earlier comment, I might just hide/remove net_pkt_alloc, net_pkt_alloc_on_iface. I thought it could be useful to somebody, I haven't found any usage for myself at least.

pfalcon reviewed (dismissed):
Just to clarify, I started to review this with the idea to have steady progress with merging @tbursztyka's net buffer refactoring changes. I see this is marked "[WIP]".

And well, indeed, I see that this is somewhat, starting with regular typos. I may imagine details of design and API may need more thinking out too.

So, let me mark that this is not ready for merging in its current shape. @tbursztyka, if you disagree, or have specific suggestions or plans to share in that regard, please let me know.

tbursztyka commented:
@pfalcon  No it's fine that it stays at -1 for the time being. There are still a lot of patches to come, and tests need to pass etc...

Commit:
net/icmpv6: Switch error message to new API

Logic does not change much, besides using new create/finalize functions
etc...

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Add support for IPv6 on the new send/sendto functions

It can now use the new IPv6 function to create IPv6 header.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv6: Make header creation function public

This will be used in ipv6_nbr, ipv6_mld and so on...

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch NA msg sending to new net_pkt API

And let's use net_ipv6_create, net_icmpv6_create, net_ipv6_finalize to
factorize the code better.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch NS msg sending to new net_pkt API

And let's use net_ipv6_create, net_icmpv6_create, net_ipv6_finalize to
factorize the code better.

De-clutter the code by reordering where src/dst are evaluated.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch RS msg sending to new net_pkt API

And let's use net_ipv6_create, net_icmpv6_create, net_ipv6_finalize to
factorize the code better.

Removing useless setup_headers private function now.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Removing useless macros in nbr code

append/append_all are unused anywhere.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv6: Improve callback signature to pass icmp header

In nbr part, this is a useful information. Since net_icmpv6_input has
already parsed the icmpv6 header, let's get pass it, instead of
retrieving/parsing it again in various handler functions.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch NA msg handler to new net_pkt API

Reworking the logic to reduce the amount of variables.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch NS msg handler to new net_pkt API

Reworking the logic to reduce the amount of variables.

Also taking the opportunity to normalize drop messages.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch RA msg handler to new net_pkt API

Reworking the logic to reduce the amount of variables.
This part was heavier to change as it was not accessing the headers
directly but instead was read parts by parts.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/icmpv6: Remove various useless get/set hdr functions

All of these are unused anywhere now, thus can be safely removed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch MLD query msg handler to new net_pkt API

Reworking the logic to reduce the amount of variables.
Introducing a generic struct to acces the common part of MLD queries,
instead of accessing part by part.

Also, returning NET_OK in case parsing went fine. We send an MLD report
anyway, so it's not a good idea to count the message as being dropped in
statistics.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch IPv6 sending preparation to new net_pkt API

Also, return a verdict instead of a pointer to net_pkt. It's simpler as
it will be up to net_send_data()'s caller to unref the net_pkt in case
of NET_DROP: less places where net_pkt can be unref.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch MLD join, leave and report to new net_pkt API

Most of the code had to be reworked due to the new API: it's more
logical to do everything sequentially (first headers, then MLD part)
than the contrary with inserting headers at the end.

Using get_data/set_data as well it makes the code clearer.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Re-enable support for fragmented packet at input

Now using the new API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Switch sending fragmented IPv6 packet to new API

This optimizes the memory quite a bit since we do not need to clone nor
split the original packet at any time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv4: Move autoconf to new net_pkt API

Only the allocator needed to be changed here.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/arp: Switch ARP to new net_pkt API

Only the allocators needed to be changed there.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/gptp: Switch GPTP to new net_pkt API

Allocation and a minor writing logic needed to be changed.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/lldp: Switch LLDP to new net_pkt API

Minor changes:
- allocator needed to be changed
- and writing the lldp_pdu as well

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Add a marker to identify LLDP message

This will be used by Ethernet L2 to set the right PTYPE in its header.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/lldp: Fix where Ethernet PTYPE is set for LLDP message

Since the rework of L2/L3 split, only L2 has access to its header. Thus
up to Ethernet one to set LLDP PTYPE.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Expose new functions to create ipv4/6 packet from context

These will be specifically needed in TCP, as well as being used in
context internally.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/tcp: Switch TCP segment preparation to new net_pkt API

Using new functions for net_context, ipv4 and ipv6 as well.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/conn: Change callback signature to pass TCP/IP headers

As these were parsed already by IPv4/6 input functions let's use them.

Applying the change on trivial UDP usage. TCP usage will have its own
commit.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/tcp: Apply connection callback signature change

This proovse to drastically reduce runtime overhead as it does not need
to parse IP nor TCP header all over again in a lot of places.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/tcp: Switch rest of TCP to new net_pkt API

Only next to be removed functions like net_tcp_set_checksum() are left
untouched. All the rest is switched.

Adding net_tcp_finalize() to follow the same logic as for UDP and else.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/icmpv6: Real packet length must be given

The test counts the '\0' of the data copied, though it does not belong
to the packet.

Since the test by-passes net_ipv6_input, which would update the packet
length according to what is in the IPv6 header, the checksum calculation
fails.

Before, checksum calculation used to grab the length from IPv6 header,
but that changed to use net_pkt_get_len(): that one must provide the
real lenth.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Enable support for TCP on new sendto function

Let's enable usage of TCP through net_context on its new sendto
function.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/tcp: Fix net_tcp_print_recv/send_info macros

Context was missing, and anyway the familiy can be grabbed from the
packet.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Add a debug option to track free pkt access more easily

Now that net_pkt are accessed through a common r/w API, using below a
net_pkt_cursor, let's have an option that will reset this cursor once
the net_pkt is freed.

Result is instead of segfaulting on r/w access, these operations will
bail out properly. Subsequent, and logical (unless you have a leak
which is another issue) net_pkt_unref will tell you who/where the pkt
was freed. Without it, you will get a segfault for instance, but that
won't tell you the exact reason. This options can help you then.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Use next_hdr field only for IPv6

Also, store the actual next_hdr value and not it's position.
This permits to reduce net_pkt from some bytes.

Such field was unused until now, but it will be soon.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ipv6: Account for extension headers when finalizing

Extension headers are specific and use nexthdr attribute from net_pkt.
next_header_proto is the user protocol (if any) that might require
finalization too (calculating the checksum).

However, if there are extension header, we need to skip those and jump
to the position where the user protocol starts.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ip: Let's make public the 2 utility unions for ip/proto headers

Though these are currently used by the core only, it will be then used
by net_context as well. This one of the steps to get rid of net_pkt's
appdata/appdatalen attributes.

Also normalizing all ip/proto parameters name to ip_hdr and proto_hdr.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Make recv_cb providing the ip and protocol headers

If status is 0, both ip_hdr and proto_hdr will own a pointer to the
relevant IP and Protocol headers. In order to know which of ipv4/ipv6
and udp/tcp one will need to use respectively net_pkt_family(pkt) and
net_context_get_ip_proto(context).

Having access to those headers directly, many callbacks will not need
to parse the packet again no get the src/dst addresses or the src/dst
ports. This will be change after this commit.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented:
ping @pfalcon, we are preparing this for a merge, could you re-visit the PR.

pfalcon commented:
I've dismissed my older review.

jukkar reviewed (approved):
LGTM
